tutorials/RandomNumberGeneration/Tests.qs

```diff
@@ -0,0 +1,152 @@
+        Message("Testing...");
+        CheckFlatDistribution(RandomTwoBits_Wrapper, 2, 1, 2, 200);
+    }
+
+    operation RandomTwoBits_Wrapper (throwaway: Int) : Int {
+        return RandomTwoBits();
+    }
+    
+    // Exercise 3.
+    operation T3_RandomNBits_Test () : Unit {
+        Message("Testing N = 1...");
+        CheckFlatDistribution(RandomNBits, 1, 0, 2, 465);
+        Message("Testing N = 2...");
+        CheckFlatDistribution(RandomNBits, 2, 1, 2, 200);
+        Message("Testing N = 10...");
+        CheckFlatDistribution(RandomNBits, 10, 461, 563, 0);
+
+    }
+
+    // f: operation to be tested
+    // numBits: number of bits in each result
+    // lowRange, highRange: median and average for generated dataset should be within [lowRange, highRange]
+    // minimum: the minimum number of times each possible number should be generated
+    operation CheckFlatDistribution (f : (Int => Int), numBits: Int, lowRange: Int, highRange: Int, minimum: Int) : Unit {
+        let max = PowI(2, numBits);
+        mutable counts = ConstantArray(max, 0);
+        mutable average = 0;
+        let total = 1000;
+
+        ResetOracleCallsCount();
+        for (i in 1..total) {
+            let val = f(numBits);
+            set average = average + val;
+            set counts w/= val <- counts[val] + 1;
+        }
+        CheckRandomCalls();
+
+        set average = average / total;
+        if (average < lowRange or average > highRange) {
+            fail $"Unexpected average. Expected between {lowRange} and {highRange}, instead was {average}";
+        }
+
+        let median = FindMedian (counts, max, total);
+        if (median < lowRange or median > highRange) {
+            fail $"Unexpected median. Expected between {lowRange} and {highRange}, instead was {median}.";
+
+        }
+
+        for (i in 0..max - 1) {
+            if (counts[i] < minimum) {
+                fail $"Unexpectedly low number of {i}'s generated. Only {counts[i]} out of {total} were {i}";
+            }
+        }
+    }
+
+    operation FindMedian (counts : Int [], arrSize : Int, sampleSize : Int) : Int {
+        mutable x = 0;
+        for (i in 0..arrSize - 1) {
+            set x = x + counts[i];
+            if (x >= sampleSize / 2) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    // Exercise 4.
+    operation T4_WeightedRandomBit_Test () : Unit {
+        ResetOracleCallsCount();
+        CheckXPercentZero(0.0);
+        CheckXPercentZero(1.0);
+        CheckXPercentZero(0.25);
+        CheckXPercentZero(0.5);
+        CheckXPercentZero(0.75);
+        CheckRandomCalls();
+    }
+
+    operation CheckXPercentZero (x : Double) : Unit {
+        Message($"Testing x = {x}...");
+        mutable oneCount = 0;
+        ResetOracleCallsCount();
+        for (N in 1..1000) {
+            set oneCount += WeightedRandomBit(x);
+        }
+        CheckRandomCalls();
+
+        let zeroCount = 1000 - oneCount;
```

Comment:
`1000` is hardcoded quite a lot in this test - it will be more convenient to define a variable `nRuns = 1000` in the beginning and refer to it throughout the code. Then if you want to change the number of runs, you change it in one place, with less risk of introducing discrepancies.